### PR TITLE
Feature : added new PlanarControls

### PR DESCRIPTION
**Pull request by EmmanuelSchmuck**
PlanarControls provides ergonomic camera controls for the planar mode : drag, pan, orbit,
zoom and animated expression to smoothly move and orient the camera.

Modified planar example to use these controls
Update contributors

### CODE REVIEW

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (peppsac) [inline src/Renderer/ThreeExtended/PlanarControls.js]: How about storing all these inside the `PlanarControls` instance so a user can instanciate several controls?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (peppsac) [inline src/Renderer/ThreeExtended/PlanarControls.js]: I don't mind the `=========` separator but this is something that should be done everywhere or nowhere (as all other stylistic choices).
So I'd suggest removing them from this PR and open a specific PR to convert itowns to this style (but I'm not sure it would be merged...)

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (peppsac) [inline src/Renderer/ThreeExtended/PlanarControls.js]: You should directly use `THREE.Math.clamp(value, 0, 1)` instead of adding a one liner function.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (peppsac) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Minor style comment:
```js
 const mouseButtons = {
    LEFTCLICK: THREE.MOUSE.LEFT,
    MIDDLECLICK: THREE.MOUSE.MIDDLE,
    RIGHTCLICK: THREE.MOUSE.RIGHT,
};
```
is prefered.

**Review by peppsac — COMMENTED** [review]
Thanks for the PR and welcome to iTowns project :)

I did a very quick first review.
I (or another reviewer) will have do a more in depth one before merging but the code is looking good mostly and is well documented, nice!

**Comment by EmmanuelSchmuck** [comment]
Thank you for the review @peppsac , I made the modifications you suggested.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (peppsac) [inline src/Renderer/ThreeExtended/PlanarControls.js]: `this.position` isn't really useful: let's use `this.camera.position` directly instead.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (peppsac) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Why don't you use the `dt` provided as an argument to your update function (see https://github.com/iTowns/itowns/blob/master/src/Core/MainLoop.js#L63)?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (peppsac) [inline src/Renderer/ThreeExtended/PlanarControls.js]: You don't need to call this: it's done for you in view's update process (https://github.com/iTowns/itowns/blob/master/src/Core/MainLoop.js#L96)

**Review by peppsac — COMMENTED** [review]
I tested this controls quickly and it seems to be working good.

I left a few comments and then it'll be ready for merging.

**Comment by EmmanuelSchmuck** [comment]
@peppsac : modified again following your suggestions. Also, my inputs no longer call update() but notifyChange() instead, much cleaner.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: `this.clock` is not used.

Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: Actually, all three "time management" vars are not used, if I'm not mistaken.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: As everything is defined in the constructor, you don't really need to attach these handlers on `this`. They can be local variables. That's better because they are truly isolated then (and being able to really isolate things is a very good reason to do this way, and not to use `prototype` or ES6 classes :-) )

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Not sure we want to document this. It's just the implementation of a FrameRequester. It's not meant to be called by the outside world. Better not have it in documentation.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: This is also true for every local state you don't want to expose. I'm thinking about:
- this.view
- this.camera
- all `travelFoo` vars
- all `drag` vars
-  this.mousePosition, this.lastMousePosition and this.deltaMousePosition
- isCtrlDown

I'm pretty sure these should be hidden.

In contrast, I believe keeping all the option properties visible is perfect :-)

For `this.state`, I have no strong opinion about it.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: What about reasoning in ms everywhere?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: p is 1.20?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: nits: switch-case?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: I believe you don't need `isCtrlDown`, use event.ctrlKey instead (available for [MouseEvent](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/ctrlKey) and [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/ctrlKey)).

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: So we call `handleRotation` at each mouse event. I have the feeling that doing all this work in `update` instead would yield the same result with better perf (actually that's one of the reason I wrote the FrameRequester logic : to give a painless mechanism to throttle input events). The logic is to set the state in event handlers, then deal with them (movements, animation etc... in `update`). 

I know `FlyControls` doesn't do things this way, but it should have done so imho (my bad). Or maybe I missed something?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Use `event.ctrlKey` instead (see below).

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: switch-case block?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: I agree this looks more practical in our examples, but please consider that one html page could have several itowns instance, or other js actions using keys... So I'll listen to `domElement` events instead.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: To elaborate: sometimes you change the position of the camera in an event (ex `handleRotation`), sometimes you change it in `update` (ex the drag logic). I believe that we should do calculations in the `update` macrotask (meaning: in the same macrotask of the `requestAnimationFrame`) as much as possible, not in the events macrotasks. So here, you'd only need to record positions.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Let's put this temp var in a closure to avoid instanciating it at each frame (by experience, creating these temp vars is very costly and the cause of every movement stuttering I had when implementing controls). We only need one per functions.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Put this temp var in a closure.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: I have the feeling this might be better in PlanarView (same thing for `getWorldPointAtScreenXY`). I'm not sure about that though. WDYT?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: > The drag movement is previously initiated when user does a left-click, by initiateDrag()

Maybe replace by

> The drag movement is previously initiated by initiateDrag()

because the fact that drag movement is done by left-click is an ergonomy detail (and could be made configurable in the future).

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Use `this.camera.position.set(x, y, z)` directly.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: You could pass directly `{x: 0.5 * window.innerWidth, y: 0.5 * window.innerHeight}` to `getWorldPointAtScreenXY`, I *think* it'll be cheaper than to create a Vector2... or modify getWorldPointAtScreenXY to take directly x and y? Caller would not need to create an object this way.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: You want the center of the canvas, not the window (remember, the canvas can be integrated in whatever webpage). You already have a ref to the canvas (`domElement`) and you'll need to use `clientHeight` and `clientWidth` (I guess). Please also change other instances of windows.innerWidth in this file.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: same, I'm pretty sure you want the canvas clientWidth.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Create the quaternion in a closure.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Should we limit the smartzoom action to a click in the plane limits?

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: 5000 seems to be a magic number... Beside, I would have expected something involving only z, not all components of the distance, can you elaborate on this calculus?

Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: Or seeing what happens below, maybe `targetHeight` could be named differently?

Btw, "I found it nicer this way" is a perfect valid answer here :-)

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Then the test should be `if (targetOrientation instanceof THREE.Quaternion)`?

Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: (it would also work if `targetOrientation` is undefined this way)

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: I would test `targetOrientation instanceOf THREE.Quaternion || targetOrientation instanceOf THREE.Vector3`, it would also cover `undefined`, `null`, whatever string and invalid objects the caller gave you. Here the caller needs to remember to pass `'none'` to explicitly disable rotation.

Not using `isVector3` because this would throw an error if undefined or null. (I don't know why Vector3 implementors felt the need to have this boolean, it's useless imho).

Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: (I assume you don't want a rotation to the default value of `travelEndRot` if `targetOrientation` is invalid)

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: nits: remove empty line (a reader could miss the `else if` below)

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: then `autoTravelTimeDist` is a factor, not a real distance? Maybe we could rename it then.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: (and you need to use canvas.clientHeight)

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Can you explain this line to me? Why isn't it always `maxAltitude`?

**Review by autra — CHANGES_REQUESTED** [review]
Taking over this, as @peppsac is off. Sorry for the inconvenience to have another reviewer with different expectations stepping in :-/

On overall I really like it! The ergonomy (and the code) is good :-) I made a lot of comments, but most are minor.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: It's not a big deal but, `dir` does not need to be declared here. You can declare it when you first use it. Anyway, `dir` and `vector` will store the same pointer, because `unproject`, `sub`, `normalize` and `multiplyScalar` all modify the vector3 inplace and returns it (see for instance [how `sub` is implemented](https://github.com/mrdoob/three.js/blob/master/src/math/Vector3.js#L161)).

On the other hand, `pointUnderCursor` **must not** be declared here, otherwise we would always return the same pointer, which is dangerous. We need to create a new instance each time, which is done by the `clone` anyway. So in short, just
```js
return this.camera.clone().add(dir.multiplyScalar(distance));
``` 
instead of line 538-540 is fine (that being said, you can keep `pointUnderCursor` if you find it clearer, just declare it with const on line 538).

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Sorry I missed that in my previous review: let's put `this.view.getPickingPositionFromDepth(posXY)` in a variable to avoid calling it 2 times.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]:
```js
this.handlePanMovement  = (() => {
    const tmp = new THREE.Vector3();
    
    return () => {
         // ....
         tmp.set(panSpeed * -1 * deltaMousePosition.x, 0, 0);
         this.camera.position.copy(this.camera.localToWorld(tmp));
         // ...
   }
})();
```
should prevent one instanciation.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: what about `quat.setFromUnitVectors(this.camera.up, new THREE.Vector3(0, 0, 1)).inverse()` and using `quat` instead of quatInverse below? It saves a `clone()`.

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: Please "cache" here:

- `new THREE.Vector3(0, 0, 1)` on line 268
- `rotationXQuaternion` and `vector` on line 279-280
- `rotationZQuaternion` on line 290

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [reply src/Renderer/ThreeExtended/PlanarControls.js]: And it's totally my fault, because I was the one to ask you about `pointUnderCursor`. Sorry :-/

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: same thing as above. Sorry, my fault :-/

File: src/Renderer/ThreeExtended/PlanarControls.js
Comment (autra) [inline src/Renderer/ThreeExtended/PlanarControls.js]: You're not using `dispatchEvent`, so you don't need the 2 lines above.

**Review by autra — CHANGES_REQUESTED** [review]
Nearly there :-)

**Review by autra — APPROVED** [review]

**Comment by autra** [comment]
Yay thanks \o/

**Comment by autra** [comment]
Nice first contrib btw !

**Comment by EmmanuelSchmuck** [comment]
Thank you :)

**Pull request merged**
